### PR TITLE
Patternfly Topology

### PR DESCRIPTION
**Pull request by jasonmadigan**
New: Adds support for Patternfly Topology to render our topology.

Screenshot 2024-10-04 at 15 55 48

### CODE REVIEW

File: src/components/PolicyTopologyPage.tsx [hunk 1]
```diff
@@ -59,16 +373,51 @@ const PolicyTopologyPage: React.FC = () => {
...
+                          zoomInCallback: action(() => {
+                            controller.getGraph().scaleBy(4 / 3);
+                          }),
+                          zoomOutCallback: action(() => {
+                            controller.getGraph().scaleBy(0.75);
+                          }),
+                          fitToScreenCallback: action(() => {
+                            controller.getGraph().fit(80);
+                          }),
+                          resetViewCallback: action(() => {
```
Comment (R-Lawton) [inline src/components/PolicyTopologyPage.tsx:407]: Do we need a reset view if we have a fit to screen? They feel very similar when using. Reset view if you have the view in half screen looks kind of funky but works perfectly for fit to screen :
Screenshot 2024-10-07 at 15 21 18

File: src/components/PolicyTopologyPage.tsx [hunk 2]
```diff
@@ -1,17 +1,292 @@
...
+    nodesep: 50,
+    ranksep: 0,
+    nodeDistance: 60,
+  });
+};
+
+const customComponentFactory = (kind: ModelKind, type: string) => {
+  const contextMenuItem = (resourceType: string, resourceName: string) => (
+    <ContextMenuItem key="go-to-resource" onClick={() => goToResource(resourceType, resourceName)}>
+      Go to Resource
```
Comment (R-Lawton) [inline src/components/PolicyTopologyPage.tsx:258]: 404 on listener go to resource

File: src/components/PolicyTopologyPage.tsx [hunk 3]
```diff
@@ -1,17 +1,292 @@
...
+// TODO: need a generic way to fetch latest versions of resources based on kind + group
+const resourceGVKMapping: { [key: string]: { group: string; version: string; kind: string } } = {
+  Gateway: { group: 'gateway.networking.k8s.io', version: 'v1', kind: 'Gateway' },
+  HTTPRoute: { group: 'gateway.networking.k8s.io', version: 'v1alpha1', kind: 'HTTPRoute' },
+  TLSPolicy: { group: 'kuadrant.io', version: 'v1alpha1', kind: 'TLSPolicy' },
+  DNSPolicy: { group: 'kuadrant.io', version: 'v1alpha1', kind: 'DNSPolicy' },
+  AuthPolicy: { group: 'kuadrant.io', version: 'v1beta2', kind: 'AuthPolicy' },
+  RateLimitPolicy: { group: 'kuadrant.io', version: 'v1beta2', kind: 'RateLimitPolicy' },
+  ConfigMap: { group: '', version: 'v1', kind: 'ConfigMap' },
+  Listener: { group: 'gateway.networking.k8s.io', version: 'v1', kind: 'Listener' },
```
[the remark this answers is not in the document]
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx:53]: Not a resource, toyed with the idea of Listener's -> their Gateway, but I think I'll just ditch the link...

File: src/components/PolicyTopologyPage.tsx [hunk 4]
```diff
@@ -1,17 +1,292 @@
...
+  action,
+  DefaultGroup,
+} from '@patternfly/react-topology';
+
+import { CubesIcon, CloudUploadAltIcon, TopologyIcon, RouteIcon } from '@patternfly/react-icons';
 import * as dot from 'graphlib-dot';
+import { kindToAbbr } from '../utils/modelUtils';
 import './kuadrant.css';
 
+// TODO: need a generic way to fetch latest versions of resources based on kind + group
```
[the remark this answers is not in the document]
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx:44]: yeah, unfortunately both expect a version to be passed too, which is what I don't want really

File: src/components/PolicyTopologyPage.tsx (hunk 1 above)
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx:407]: had a similar thought earlier on, and this reminds me of it. I think fit is a better "reset" anyway

File: src/components/PolicyTopologyPage.tsx (hunk 2 above)
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx:269]: fixed, goes to the Gateway now

File: src/components/PolicyTopologyPage.tsx (hunk 1 above)
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx]: removed, just fit as an action now

File: src/components/PolicyTopologyPage.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasonmadigan) [reply src/components/PolicyTopologyPage.tsx:53]: fixed. Listeners now have a special case, navigates to associated GW.

**Comment by jasonmadigan** [comment]
@R-Lawton fixed up

File: src/components/PolicyTopologyPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (R-Lawton) [reply src/components/PolicyTopologyPage.tsx:44]: ah ok thats annoying

**Comment by R-Lawton** [comment]
Noticed one small thing now with the change to listener links. They text boxes overlap each other with which ever you hover over being in the front
Screenshot 2024-10-07 at 16 47 40

**Comment by jasonmadigan** [comment]
Added a small fix, although with how the component works there may be cases where labels can unavoidably overlap. There are options to prevent node overlap, sadly none really for labels

**Review by R-Lawton — APPROVED** [review]
Looks great nice work 🎉

**Pull request merged**